src/hotspot/share/gc/shenandoah/shenandoahForwarding.inline.hpp

```diff
@@ -75,13 +75,18 @@ inline bool ShenandoahForwarding::is_forwarded(oop obj) {
 }
 
 inline oop ShenandoahForwarding::try_update_forwardee(oop obj, oop update) {
-  markWord old_mark = obj->mark();
+  markWord old_mark = obj->mark_acquire();
   if (old_mark.is_marked()) {
     return oop(old_mark.clear_lock_bits().to_pointer());
   }
 
   markWord new_mark = markWord::encode_pointer_as_mark(update);
-  markWord prev_mark = obj->cas_set_mark(new_mark, old_mark, memory_order_conservative);
+
+  // We need a stronger acq/rel memory order here. Release on successful forwardee
+  // update guarantees publishing the new object copy contents correctly
+  // (would be paired with acquire in forwardee accessors). Acquire on failed update
+  // would get the updated object after the forwardee load.
+  markWord prev_mark = obj->cas_set_mark(new_mark, old_mark, memory_order_acq_rel);
```

Comment:
You have obj->mark_acquire() above, I don't think you need leading acquire here.

Reply:
A bit different: we want to acquire `prev_mark` (failure witness) here, if we lose the update race. So the `old_mark` acquire does not really help us.